### ISSUE TITLE
`Lint/AmbiguousOperatorPrecedence` configuration

### ISSUE THREAD

**Issue by pyromaniac** [rubocop/rubocop#10080]
## Is your feature request related to a problem? Please describe.

The new `Lint/AmbiguousOperatorPrecedence` cop looks way like an overkill. Parenthesis in arithmetic expressions for all the operations create a lot of visual noise and might be not a god option for a lot of developers. I believe, it is going to be disabled by this reason a lot. 

But for boolean expressions additional parenthesis could bring a lot of value.

## Describe the solution you'd like

I believe we need to split operators into groups. I see two right now: boolean and arithmetic. Would be great to be able enabling them separately.

## Describe alternatives you've considered

The only alternative for now is to disable the cop completely.

**Comment by dvandersluis** [comment]
Thank you for your feedback. The idea is while the cop is pending to gauge how well it works for RuboCop users and tweak as necessary.

The cop is intended to make precedence very explicit, even if you're not very familiar with Ruby's order of operations, and as such it may add more parentheses that you might expect, or that seem excessive even, depending on what your original code was. That might also be an indication that there's a useful refactoring there.

Being able to disable groups as you suggested is interesting, but has some pitfalls. First off, operations can be overridden, so `+` isn't necessarily arithmetic, for example. The shift operators (`<<` and `>>`) are a good example of how operators that are commonly overridden to mean different things. Secondly, I'm not sure that the other bitwise operators (`&`, `|`, `^`) cleanly fit into "arithmetic" either.

@bbatsov what do you think?

**Comment by bbatsov** [comment]
I've also been thinking about some form of configuration, but I don't think that the idea about groups will work in practice. We can probably have a list of operators on which cop would apply that could narrow its scope, but I'm not eager to introduce this until we've gathered more feedback from users.

**Comment by januszm** [comment]
it reports false positives, e.g.:

```
W: [Correctable] Lint/AmbiguousOperatorPrecedence: Wrap expressions with varying precedence with parentheses to avoid ambiguity.

...
something =  31 * 24 + 10 * 10
```

there is no ambiguity in this example.

**Comment by bbatsov** [comment]
@januszm There's never any ambiguity with respect to operator precedence for Ruby itself. We used the word in the sense that there's potential ambiguity for the readers of the code (at the very least they have to read the code more carefully). Parentheses eliminate this.

**Comment by bendilley** [comment]
If it were up to me, this cop would be called `Style/ExplicitOperatorPrecedence` and disabled by default. Having done plenty of algebra at school, I cannot see any ambiguity, nor even delayed comprehension, in the precedence of arithmetic operators, and the same goes for anyone familiar with formal logic looking at the boolean operators.

If I liked parentheses everywhere I wouldn't be using Ruby :trollface:  Seriously though, I can see a lot of people liking this cop too. I just feel it belongs under Style, not Lint.

Regarding the OP's grouping suggestion, perhaps there could be a configuration option in which users can define which operators they'd like included/excluded, with all included by default?

**Comment by bendilley** [comment]
On further reflection, where I do think this would have a lot of value as a Lint cop is where there is a mixture of different operator groups, e.g. arithmetic operators mixed with boolean, or (and this would be easy to do by accident) boolean with bitwise operators. In those cases I would definitely have to [look up](https://ruby-doc.org/core/doc/syntax/precedence_rdoc.html) the operator precedence, so parentheses or a rethink would then be an uncontroversial suggestion by rubocop.

Again, perhaps the user could be allowed to define explicitly a list of groups of operators they don't mind mixing together?

**Comment by johnnyshields** [comment]
I agree with the idea of configurable ambiguity cases which can be toggled on/off. For my tastes, there is no ambiguity in `3 * 5 + 4`

**Comment by dvandersluis** [comment]
> I believe, it is going to be disabled by this reason a lot.

> If it were up to me, this cop would be ... disabled by default

I am generally curious if RuboCop users have a problem with disabling a cop in your own configuration that doesn't suit your purposes, and if so, why?

**Comment by nevans** [comment]
on topic for ticket: I agree that this rule should be configurable.  I understand why it's in `Lint`; that rationale makes sense.  But IMO it belongs in `Style/ExplicitOperatorPrecedence`.

tangential to ticket, but responding (verbosely) to:
> I am generally curious if RuboCop users have a problem with disabling a cop in your own configuration that doesn't suit your purposes, and if so, why?

**tl;dr: `Lint` rules are special.**

I "love" disabling rules.  I keep all rules on by default, so every time I upgrade I _get to disable a bunch of new ones!_  I get to argue about how these crazy rubocop rules are wrong and why can't everyone see that the correct way is my wa... _devolves into an hour of ranting about the importance of properly deployed whitespace_  😜  I also learn about new techniques I hadn't previously known about.  And I get to reconsider some things I'd been sloppy with.  And my my legacy code dating all the way back to 1.8.6 gets a nice auto-upgrade to features in the latest versions of ruby.

But seriously, I have no problem disabling rules.  It's no hardship.  My default `.rubocop.yml` is ~418 lines, most of those lines are disabled rules, and many of them are commented, e.g. "if rubocop supported such-and-such style I would love this rule. but it doesn't so it's the worst and I hate it. maybe I should learn how it works so I can submit a PR".  Even with comments, I waste no time tossing new lines into my config:
````yaml
Style/CalledPrettierButActuallyVeryUgly: { Enabled: false } # UGH. WHY?! You had ONE JOB!
Style/ThisContentiousRuleIsNowAStandard: { Enabled: false } # is this an official standards body? 😦👎🏻 
Style/WearPantsOnHead: { Enabled: false } # what were they thinking!?!
````

I'm being silly here, but I really _do_ enjoy going through the new batch of rules every time I upgrade.  I always learn something new.  And I always get the opportunity to auto-correct something marvelous.  And my arguments with rubocop are in good fun, unlike my arguments with prettier... those never end well (prettier is banned from my projects).  My approach adds extra time to upgrades, so I don't always stay up-to-date with the latest versions.  But I prefer it this way: It's like I get to have a spirited conversation with the rubocop developers after every upgrade.

When people or teams disagree with as many defaults as I do, they are often advised to start with everything disabled by default and only enable the ones they like.  That's a valid approach and good advice, IMO.  But `Lint` rules are the ones that I advise everyone always start with enabled by default.

I try to give the `Lint` rules special reverence.  I pause long and think hard before disabling them.  They highlight and catch bugs.  I'll deal with ugly code if it catches bugs.  I'll live with lint rules that I hate for weeks or more to see if I'll come around.  I want to give them the time to persuade me and sometimes they do.  In my 418 line `.ruboconf.yaml` file I've only disabled four `Lint` rules, including this one. (I have a couple dozen in the todo file, so maybe I'd disable a couple of those too if I'm forced to confront them.)

I hope that makes sense. **tl;dr: `Lint` rules are special and should be special IMO.  I _do_ hesitate to disable them.**  Should I change my strategy?

**Comment by JochenLutz** [comment]
While I came here to complain about unnecessary noise in arithmetic expressions (which I did with a :+1: in the first comment), I want to add an example it found in our code base why it should stay in `Lint`:

`array_var << model.attribute_which_might_be_nil || '-'`

This should probably add a value with fallback to an array, but the rewritten code shows that it does something else:

`(array_var << model.attribute_which_might_be_nil) || '-'`

**Comment by jaredbeck** [comment]
> .. perhaps there could be a configuration option in which users can define exactly which operators they'd like included/excluded, with all included by default?

:+1: but with arithmetic operators excluded by default

**Comment by JochenLutz** [comment]
The issue still persists.

**Comment by reachire-smendola** [comment]
I agree with those that said that operator precedence for purely arithmetic operators is unambiguous to everyone.

There are operators whose precedence is not immediately obvious to a coder; e.g. in JS it might be one way, in ruby another, and in python yet another, and it's hard to remember what's what. Also some operators may have unintuitive grouping regardless of any other language. But + - * / are not among them.

As it is, I can't use this cop because I neither want to see this warning on 2 * n + 1, nor do I want to add parentheses to that expression.  But I would appreciate having it for more complex expressions especially involving ?:

**Comment by forthrin** [comment]
This is how AI draws the line in a potential liberalization of the cop in question.

### Most intention aligned

| Expression | Dev hubris | Matz blindside |
|-|-|-|
| x = a or b | x = (a or b) | (x = a) or b |
| x = a and b | x = (a and b) | (x = a) and b |
| -a ** 2 | (-a) ** 2 | -(a ** 2) |
| a + b * c | (a + b) * c | a + (b * c) |
| a << b + c | (a << b) + c | a << (b + c) |
| a & b == c | a & (b == c) | (a & b) == c |
| a == b && c == d | a == (b && c) == d | (a == b) && (c == d) |
| a && b \|\| c | a && (b \|\| c) | (a && b) \|\| c |
| a \|\| b ? c : d | a \|\| (b ? c : d) | (a \|\| b) ? c : d |
| a..b == c | a..(b == c) | (a..b) == c |
| a ? b : c \|\| d | (a ? b : c) \|\| d | a ? b : (c \|\| d) |

### Candidates for clemency

```ruby
# Accept: multiplication/division binds before addition/subtraction
_value = a * b - c / d

# Accept: && binds more tightly than ||
_allowed = authenticated && active || admin
_allowed = public_user || subscribed && verified

# Accept: arithmetic binds more tightly than shifts
_value = 1 << width + offset
_value = offset + index << bits

# Accept: exponentiation binds more tightly than multiplication
_value = base ** exponent * scale

# Accept: arithmetic binds more tightly than range operators
_range = start + offset..finish + offset
```
